keps/sig-network/20191020-default-deny-pod-traffic

```diff
@@ -0,0 +1,270 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Make pod-to-pod communication in Kubernetes secure by default.
+Pods should require an explicit NetworkPolicy ingress configuration,
+in order to receive traffic.
+
+This proposal explicitly targets ingress traffic.
+Egress traffic is not _currently_ being targeted,
+not is not yet deemed out of scope.
+
+## Motivation
+
+Kubernetes and Kubernetes workloads are notoriously easy to leave insecure.
+Many users are not aware of relevant settings (like NetworkPolicy),
+or choose not to configure them due to friction and learning curves.
+This is a systemic pattern,
+NetworkPolicy is one example.
+
+TODO pull examples
+
+### Goals
+
+* Restrict pod ingress by default on a cluster.
+
+### Non-Goals
+
+What is out of scope for this KEP?
+Listing non-goals helps to focus discussion and make progress.
+
+## Proposal
+
+There are multiple options for design/implementation.
+
+Restricting network access requires a CNI driver that supports NetworkPolict.
+All options will require codifying NetworkPolicy support as a requirement,
+or creating a clear user experience when using a CNI driver that does not support NetworkPolicy.
+
+### Option 1: Change default NetworkPolicy behavior
+
+The most conceptually simple option is to change default NetworkPolicy behavior.
+Currently,
+NetworkPolicy allows all traffic to pods that are not selected by a NetworkPolicy. [1]
+
+If all pods are treated as isolated,
+users must explicitly create NetworkPolicy objects
+(either as needed, or by allowing all ingress and _explicitly_ enabling the existing behavior).
+This would potentially require a v2 API, due to changes in behavior (if not the API itself).
+
+### Option 2: Inject explicit deny-all NetworkPolicy objects
+
+If making changes to NetworkPolicy API is too undesirable,
+we could automatically apply NetworkPolicy objects to achieve the same effect.
+
+There is no precedent(?) or mechanism(?) for performing pre-creation actions
+(IE the opposite of finalizers).
+
+A simple controller could monitor for new namespaces,
+but it would be subject to race conditions,
+where new pods would be subject to the default-open behavior before the NetworkPolicy is created.
+
+### Risks and Mitigations
+
+Changing network defaults will existing break user setups (EG apps with no NetworkPolicies),
+and may break existing clusters if designed or applied incorrectly.
+If this behavior is introduced to an existing cluster,
+it should be done in a permissive fashion
+(EG marking the cluster or specific namespaces as using "legacy mode").
+
+This change may frustrate users,
+as it introduces another step (configure NetworkPolicy) that was previously optional.
```

Comment:
No, because networkpolicy is having race condition since forever and there is no way of fixing it in the current state.
Pod ready++ was more a joke, than a fix and tear down is not even thought about

Reply:
In general, Kubernetes currently leans towards making things easy for users who don't need to worry about security and assuming that people who do need to worry about security will... worry about security, and end up doing the right thing...

Reply:
I'm all in favor of making things more correct by default, but @szuecs this is not a helpful comment.

Reply:
@thockin , sorry, let's ignore the mentioned problem.